pkg/kv/dynamodb/store.go

```diff
@@ -0,0 +1,361 @@
+	}
+
+	item := DynKVItem{
+		PartitionKey: string(partitionKey),
+		ItemKey:      string(key),
+		ItemValue:    string(value),
+	}
+
+	marshaledItem, err := dynamodbattribute.MarshalMap(item)
+	if err != nil {
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+
+	input := &dynamodb.PutItemInput{
+		Item:      marshaledItem,
+		TableName: &s.params.TableName,
+	}
+	if usePredicate {
+		if valuePredicate != nil {
+			input.ConditionExpression = aws.String(ItemValue + " = :predicate")
+			input.ExpressionAttributeValues = map[string]*dynamodb.AttributeValue{
+				":predicate": {S: aws.String(string(valuePredicate.([]byte)))},
+			}
+		} else {
+			input.ConditionExpression = aws.String("attribute_not_exists(" + ItemValue + ")")
+		}
+	}
+
+	_, err = s.svc.PutItemWithContext(ctx, input)
+	if err != nil {
+		if _, ok := err.(*dynamodb.ConditionalCheckFailedException); ok && usePredicate {
+			return kv.ErrPredicateFailed
+		}
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+	return nil
+}
+
+func (s *Store) Delete(ctx context.Context, partitionKey, key []byte) error {
+	if len(partitionKey) == 0 {
+		return kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+
+	_, err := s.svc.DeleteItemWithContext(ctx, &dynamodb.DeleteItemInput{
+		TableName: aws.String(s.params.TableName),
+		Key:       s.bytesKeyToDynamoKey(partitionKey, key),
+	})
+	if err != nil {
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+	return nil
+}
+
+func (s *Store) Scan(ctx context.Context, partitionKey, start []byte) (kv.EntriesIterator, error) {
+	internalIter, err := s.scanInternal(ctx, partitionKey, start, nil)
+	if err != nil {
+		return nil, err
+	}
+	return internalIter, nil
+}
+
+func (s *Store) scanInternal(ctx context.Context, partitionKey, scanKey []byte, exclusiveStartKey map[string]*dynamodb.AttributeValue) (*EntriesIterator, error) {
+	if len(partitionKey) == 0 {
+		return nil, kv.ErrMissingPartitionKey
+	}
+	keyConditionExpression := PartitionKey + " = :partitionkey"
+	expressionAttributeValues := map[string]*dynamodb.AttributeValue{
+		":partitionkey": {
+			S: aws.String(string(partitionKey)),
+		},
+	}
+	if len(scanKey) > 0 {
+		keyConditionExpression += " AND " + ItemKey + " >= :fromkey"
+		expressionAttributeValues[":fromkey"] = &dynamodb.AttributeValue{
+			S: aws.String(string(scanKey)),
+		}
+	}
+	queryInput := &dynamodb.QueryInput{
+		TableName:                 aws.String(s.params.TableName),
+		KeyConditionExpression:    aws.String(keyConditionExpression),
+		ExpressionAttributeValues: expressionAttributeValues,
+		ConsistentRead:            aws.Bool(true),
+		ScanIndexForward:          aws.Bool(true),
+		ExclusiveStartKey:         exclusiveStartKey,
+	}
+	queryOutput, err := s.svc.QueryWithContext(ctx, queryInput)
+	if err != nil {
+		return nil, fmt.Errorf("%s (start=%v): %w ", err, string(scanKey), kv.ErrOperationFailed)
+	}
+
+	return &EntriesIterator{
+		store:        s,
```

Comment:
Save the context - and use it in the tmpIter as well?

Reply:
Excellent point. Done